dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/trackedentity/TrackedEntityServiceTest.java

```diff
@@ -1578,6 +1579,184 @@ void shouldReturnAllEntitiesWhenSuperuserAndModeAll()
+    makeProgramMetadataInaccessible(programC);
+    Program eventProgram = createProgramWithoutRegistration('E');
+    manager.save(eventProgram, false);
+
+    injectSecurityContextUser(user);
+    ForbiddenException exception =
+        assertThrows(
+            ForbiddenException.class,
+            () ->
+                trackedEntityService.getTrackedEntity(
+                    trackedEntityA.getUid(), null, TrackedEntityParams.TRUE, false));
+
+    assertContains(
+        String.format("User has no access to TrackedEntity:%s", trackedEntityA.getUid()),
+        exception.getMessage());
+  }
+
+  @Test
+  void shouldFailWhenRequestingSingleTEAndTETDoesNotMatchAnyProgram() {
+    injectSecurityContextUser(getAdminUser());
+    TrackedEntityType trackedEntityType = createTrackedEntityType('T');
+    manager.save(trackedEntityType, false);
+    programA.setTrackedEntityType(trackedEntityType);
+    manager.update(programA);
+    programB.setTrackedEntityType(trackedEntityType);
+    manager.update(programB);
+    programC.setTrackedEntityType(trackedEntityType);
+    manager.update(programC);
+
+    injectSecurityContextUser(user);
+    ForbiddenException exception =
+        assertThrows(
+            ForbiddenException.class,
+            () ->
+                trackedEntityService.getTrackedEntity(
+                    trackedEntityA.getUid(), null, TrackedEntityParams.TRUE, false));
+
+    assertContains(
+        String.format("User has no access to TrackedEntity:%s", trackedEntityA.getUid()),
+        exception.getMessage());
+  }
+
+  @Test
+  void shouldFailWhenRequestingSingleTEAndNoDataAccessToAnyProgram() {
+    injectSecurityContextUser(getAdminUser());
+    makeProgramMetadataAccessibleOnly(programA);
+    makeProgramMetadataAccessibleOnly(programB);
+    makeProgramMetadataAccessibleOnly(programC);
+
+    injectSecurityContextUser(user);
+    ForbiddenException exception =
+        assertThrows(
+            ForbiddenException.class,
+            () ->
+                trackedEntityService.getTrackedEntity(
+                    trackedEntityA.getUid(), null, TrackedEntityParams.TRUE, false));
+
+    assertContains(
+        String.format("User has no access to TrackedEntity:%s", trackedEntityA.getUid()),
+        exception.getMessage());
+  }
+
+  @Test
+  void shouldFailWhenRequestingSingleTEAndNoAccessToTET() {
+    injectSecurityContextUser(getAdminUser());
+    trackedEntityTypeA.setSharing(Sharing.builder().publicAccess("rw------").build());
+    manager.update(trackedEntityA);
+
+    injectSecurityContextUser(user);
+    ForbiddenException exception =
+        assertThrows(
+            ForbiddenException.class,
+            () ->
+                trackedEntityService.getTrackedEntity(
+                    trackedEntityA.getUid(), null, TrackedEntityParams.TRUE, false));
+
+    assertContains(
+        String.format("User has no access to TrackedEntity:%s", trackedEntityA.getUid()),
```

Comment:
Maybe a space in between `:` and the UID would make the error more readable. Also seems like you are doing a full comparison in which case assertEquals is what you want. assertContains is nice if you can assert on only a bit of the error message so you can change the message without having to change the test as well.

Reply:
Changed to `assertEquals`.
The space before the UID I didn't want to add because this API response is used in other places and I'm not sure whether there are clients relying on this error message in its current format.